backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/S3BackupStore.java

```diff
@@ -0,0 +1,97 @@
+import io.camunda.zeebe.backup.api.BackupStore;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+
+/**
+ * {@link BackupStore} for S3. Stores all backups in a given bucket.
+ *
+ * <p>All created object keys are prefixed by the {@link BackupIdentifier}, with the following
+ * scheme: {@code checkpointId/partitionId/nodeId}
+ *
+ * <p>Each backup contains:
+ *
+ * <ol>
+ *   <li>A 'metadata' object, containing {@link Metadata} serialized as JSON, for example
+ *       <pre>checkpointId/partitionId/nodeId/metadata</pre>
+ *   <li>A 'status' object, containing the {@link BackupStatus}, for example
+ *       <pre>checkpointId/partitionId/nodeId/status</pre>
+ *   <li>Objects for snapshot files, additionally prefixed with 'snapshot', for example
+ *       <pre>checkpointId/partitionId/nodeId/snapshots/snapshot-file-1</pre>
+ *   <li>Objects for segment files, additionally prefixed with 'segments', for example
+ *       <pre>checkpointId/partitionId/nodeId/segments/segment-file-1</pre>
+ * </ol>
+ */
+public final class S3BackupStore implements BackupStore {
+
+  static final ObjectMapper MAPPER = new ObjectMapper();
+
+  private final S3Config config;
+  private final S3AsyncClient client;
+
+  public S3BackupStore(final S3Config config, final S3AsyncClient client) {
+    this.config = config;
+    this.client = client;
+  }
+
+  @Override
+  public CompletableFuture<Void> save(final Backup backup) {
+    final var metadata = saveMetadata(backup);
+
+    return CompletableFuture.allOf(metadata);
+  }
+
+  @Override
+  public CompletableFuture<BackupStatus> getStatus(final BackupIdentifier id) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public CompletableFuture<Void> delete(final BackupIdentifier id) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public CompletableFuture<Backup> restore(final BackupIdentifier id) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public CompletableFuture<Void> markFailed(final BackupIdentifier id) {
+    throw new UnsupportedOperationException();
+  }
+
+  private CompletableFuture<PutObjectResponse> saveMetadata(Backup backup) {
+    try {
+      return client.putObject(
+          request ->
+              request
+                  .bucket(config.bucketName())
+                  .key(objectPrefix(backup.id()) + Metadata.OBJECT_KEY)
+                  .contentType("application/json")
+                  .build(),
+          AsyncRequestBody.fromBytes(MAPPER.writeValueAsBytes(Metadata.of(backup))));
+    } catch (JsonProcessingException e) {
+      return CompletableFuture.failedFuture(e);
+    }
+  }
+
+  private static String objectPrefix(BackupIdentifier id) {
+    return "%s/%s/%s/".formatted(id.checkpointId(), id.partitionId(), id.nodeId());
```

Comment:
:wrench: If I remember correctly, we decided to put in this order partitionId/checkpointId/nodeid

Reply:
Oops, good catch :+1: